### DISCUSSION TITLE
Combining repos

### DISCUSSION THREAD

**Discussion by IlyaSemenov** [romeerez/orchid-orm#101] [Ideas]
How is user supposed to combine different repos? Such as what if they add `userSearchRepo` then `userFilterRepo` then `userDeleteRepo`, but then want to use all of them in one query?

```ts
// Is that the supposed way?
userSearchRepo(userFilterRepo(userDeleteRepo(db.user))).mysearch().myfilter().mydelete()

// What if that was possible?
db.user.use(userSearchRepo, userFilterRepo, userDeleteRepo).mysearch().myfilter().mydelete()
// or
db.user.withRepos(...)
```

**Comment by romeerez** [comment]
A method for this could be added, but I doubt if this is a realistic case. In basic cases, tables have no repo, then to de-duplicated queries, it's one repo per table. If there is some large complex feature, as it was a search in my case, because it was very tricky with various `UNION`'s, full-text search, fuzzy-search, searching by relations, and so on, it makes sense to keep this stuff separate. But IMO having many different-purposed repos for a single table is too much, though maybe there are such cases.

In what you're doing, are there too much different queries to keep in a single repo?

**Reply by IlyaSemenov** [reply]
I have a set of authenticating repos (one per table, generated with a helper) and for some tables I also have their own dedicated repos.  That gives a reasonable case with at least two repos per query.

My scenario would be as follows:

```ts
// userRepo = custom user repo
// restrictedDb.user = authenticating repo
userRepo(restrictedDb.user).withRole(role).find(id).mydelete() // naming simplified for clarity
```

If wrapper composition could be replaced with `.use`-style composition, that would streamline the code:

```ts
restrictedDb.user.withRole(role).use(userRepo).find(id).mydelete()
```

That said, the selling point of repos is decoupling. In your very examples, `userRepo` is decoupled from `followRepo`. It's explained how to use `followRepo` from within `userRepo`, but isn't that expectable that users will also want to use `userRepo` and `followRepo` together in one-off queries?

**Reply by romeerez** [reply]
Yeah, great idea to move the auth methods into a repo. In this way, various integrations for auth libraries could work with the Orchid.

`use` is too generic, how about:

```ts
db.user.repo(oneRepo, secondRepo).doOne().doSecond()

// or
db.user.useRepo(oneRepo, secondRepo).doOne().doSecond()

// or this seems interesting
// word `and` would be better, but it's taken
oneRepo.mix(secondRepo).doOne().doSecond()

// mix is also too generic
// mixRepo maybe?
oneRepo(db.user).mixRepo(secondRepo, thirdRepo)
```

I'd vote for `useRepo`.

Passing `db.user` into the repo isn't required, only if this happens inside nested select, like when querying comments to select their author and use a repo for this.

**Reply by IlyaSemenov** [reply]
`useRepo` sounds good, agreed.
